### ISSUE TITLE
Feature Request: Out of Sample Predictions

### ISSUE THREAD

**Issue by juanitorduz** [bambinos/bambi#356]
Hi! By looking into the documentation I am not sure if we can get out-of-sample predictions (posterior) from a fitted model, for example like described in https://github.com/pymc-devs/pymc-examples/blob/main/examples/generalized_linear_models/GLM-out-of-sample-predictions.ipynb . 
This would be a nice feature to have. Are there any plans for it (or is already possible)? 
Thanks :)

**Comment by tomicapretto** [comment]
Hi @juanitorduz and thanks for opening the issue!

We don't have an implementation in Bambi right now, but _I think_ it is still doable. You can access the underlying PyMC3 model with `Model.backend.model`, and work as you do in the notebook. I'm not sure how well this works if you have a data set with categorical predictors.

There is another approach that requires more work, but I think it is the one we will follow for our implementation in Bambi. The object we use to represent design matrices has a method [_evaluate_new_data](https://github.com/bambinos/formulae/blob/master/formulae/design_matrices.py#L164) that allows you to obtain a new design matrix constructed from the new data, but respecting constraints specified with the original data (like the encoding used for categorical variables or the mean that was used in a call to `center(x)`).

There is an unmerged example using this (yet undocumented) feature, [here](https://github.com/bambinos/bambi/blob/1282a957e602b78237818406a018724bb9ba82cc/docs/notebooks/negative_binomial.ipynb). 
![image](https://user-images.githubusercontent.com/25507629/118485075-520d3100-b6ee-11eb-99bb-18e9a3bcee6c.png)

To summarize, it's not so easy, but it is possible, and we are definitely adding it to Bambi. I'm still not sure when, because I was told there was something being done on PyMC3 side in this respect. As I am aware, they are preparing the V4 of PyMC3 and we are waiting for that release to see what work is required to have this predict feature. 

If you have any other questions, just leave them here :)

**Comment by juanitorduz** [comment]
Thank you very much for the information! :)

**Comment by tomicapretto** [comment]
Hi @juanitorduz! This feature is coming in #371. Suggestions are much welcomed :D

**Comment by juanitorduz** [comment]
> Hi @juanitorduz! This feature is coming in #371. Suggestions are much welcomed :D

That is awesome! Thank you all!

**Comment by tomicapretto** [comment]
#372 has been merged. Thanks for opening the issue @juanitorduz :D

**Comment by AniMadurkar** [comment]
@tomicapretto so interestingly enough, "Danceability" is a numerical field! I did have a categorical field which I OneHotEncoded, but I purposely didn't include it in this dataset just to iteratively add complexity. All the values in the "train" and "test" dataframes are numerical and standardized as well.

either way, appreciate you opening an issue on this! Keep up the great work, I'm having lots of fun learning this.

**Comment by tomicapretto** [comment]
@AniMadurkar but since you are doing `(1|Danceability)` you are saying you want a varying intercept for each level in `Danceability`, which treats `Danceability` as a categorical variable.

Do you want to tell me more what you're trying to do with the group-specific intercepts? I'm afraid that using all numerical variables as groups may not be the right approach here. If  `Danceability` is a categorical variable and you want varying intercepts for each group, simply use it as it is (categorical) and Bambi will handle it appropriately.
